cura/CuraApplication.py

```diff
@@ -723,6 +738,20 @@ def preRun(self):
     def run(self):
         self.preRun()
 
+        container_registry = ContainerRegistry.getInstance()
+        self._variant_manager = VariantManager(container_registry)
+        self._variant_manager.initialize()
+
+        from cura.Machines.MaterialManager import MaterialManager
+        self._material_manager = MaterialManager(container_registry, parent = self)
+        self._material_manager.initialize()
+
+        from cura.Machines.QualityManager import QualityManager
+        self._quality_manager = QualityManager(container_registry, parent = self)
+        self._quality_manager.initialize()
+
+        self._machine_manager = MachineManager(self)
```

Comment:
Those constructor calls were actually intended. With `getSomeManager()` to create a singleton on demand, we had encountered several implicit circular dependency issue and some other issues due to unpredictable ordering of when each singleton gets initialized, so here, we plan to create and initialize singletons explicit in a pre-defined order, and the `getABC()`s will just be functions to get a singleton instance. This way it is clear what is initialized first and which dependencies a component (possibly) has.

Reply:
For the `self` part, that's the `parent` thingy for Qt. This is set because we found that if we pass an Qt object which has no parent to QML, then QML will takeover the ownership by setting the parent to the QML module that first acquires it. However, QML does smart things with garbage collection, so those objects will be deleted at some point. So, to prevent this, we explicitly added the `Application` or `CuraApplication` to be the parent of those singleton objects. This is also done in Uranium for `ContainerStack`s.